File: csharp/Othello/Othello.cs
//==========================================================
// Class Othello
// Play Othello (Reversi) on the command line
// https://en.wikipedia.org/wiki/Reversi
// Akseli Lukkarila
// 2019-2026
//==========================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace Othello;

/// <summary>Gameplay loop and main logic.</summary>
internal sealed class Othello {
    private Board _board;
    private readonly Settings _settings;
    private readonly Player _playerBlack;
    private readonly Player _playerWhite;
    private readonly List<string> _gameLog = [];
    private int _gamesPlayed;
    private int _roundsPlayed;

    /// <summary>Initialize Othello game.</summary>
    public Othello(Settings settings) {
        _board = new Board(settings.BoardSize);
        _settings = settings;
        _playerBlack = Player.Black(settings.ToPlayerSettings());
        _playerWhite = Player.White(settings.ToPlayerSettings());
    }

    /// <summary>Play one full game of Othello.</summary>
    public void Play() {
        while (true) {
            InitGame();
            GameLoop();
            PrintResult();
            if (_settings.ShowLog) {
                PrintLog();
            }
            if (_settings.AutoplayMode || !GetAnswer("Would you like to play again")) {
                break;
            }
        }
    }

    /// <summary>Initialize game board and players for a new game.</summary>
    private void InitGame() {
        // Re-use existing objects instead of initializing new ones
        if (_gamesPlayed > 0) {
            _board = new Board(_settings.BoardSize);
            _playerBlack.Reset();
            _playerWhite.Reset();
            _roundsPlayed = 0;
            _gameLog.Clear();
        }
        if (_settings.AutoplayMode) {
            // Computer plays both
            _playerBlack.SetComputer();
            _playerWhite.SetComputer();
        } else if (_settings.UseDefaults) {
            // Default: play as black against white computer player
            _playerWhite.SetComputer();
        } else if (GetAnswer("Would you like to play against the computer")) {
            if (GetAnswer("Would you like to play as black or white", "b", "w")) {
                _playerWhite.SetComputer();
            } else {
                _playerBlack.SetComputer();
            }
        }
        if (!_settings.CheckMode) {
            ColorPrint.PrintBold("\nPlayers:");
            PrintStatus();
        }
    }

    /// <summary>Keep making moves until both players can't make a move any more.</summary>
    private void GameLoop() {
        while (_board.CanPlay() && (_playerBlack.CanPlay || _playerWhite.CanPlay)) {
            ++_roundsPlayed;
            PrintRoundHeader();
            foreach (Player player in new[] { _playerBlack, _playerWhite }) {
                string? result = player.PlayOneMove(_board);
                if (result != null) {
                    _gameLog.Add($"{result};{_board.LogEntry()}");
                }
                if (!_settings.CheckMode) {
                    Console.WriteLine("--------------------------------");
                }
            }
        }
        ++_gamesPlayed;
        PrintGameEndFooter();
    }

    /// <summary>Format game log with line numbers for each move.</summary>
    private string FormatGameLog() {
        return string.Join(
            "\n",
            _gameLog.Select((line, index) => $"{index + 1:00}: {line}")
        );
    }

    /// <summary>Print header for the current round.</summary>
    private void PrintRoundHeader() {
        if (!_settings.CheckMode) {
            ColorPrint.PrintBold($"\n=========== ROUND: {_roundsPlayed} ===========");
        }
    }

    /// <summary>Print footer after the game has ended.</summary>
    private void PrintGameEndFooter() {
        if (!_settings.CheckMode) {
            ColorPrint.PrintBold("\n================================");
            ColorPrint.PrintGreenBold("The game is finished!\n");
        }
    }

    /// <summary>Print game log which shows all moves made and the game board state after each move.</summary>
    private void PrintLog() {
        string formattedLog = FormatGameLog();
        if (!_settings.CheckMode) {
            ColorPrint.PrintYellowBold("Game log:");
            Console.WriteLine(formattedLog);
        }
        string hexHash = Utils.CalculateSha256(formattedLog);
        Console.WriteLine(hexHash);
    }

    /// <summary>Print ending status and winner info.</summary>
    private void PrintResult() {
        ColorPrint.PrintBold("Result:");
        PrintStatus();
        Console.WriteLine("");

        Disk winner = _board.Result();
        Console.WriteLine(
            winner == Disk.Empty
                ? "The game ended in a tie...\n"
                : $"The winner is {winner.DiskString()}!\n"
        );
    }

    /// <summary>Print current board and player info.</summary>
    private void PrintStatus() {
        Console.WriteLine(_playerBlack);
        Console.WriteLine($"{_playerWhite}\n");
        Console.WriteLine(_board);
    }

    /// <summary>Ask a question with two options, and return bool from user answer.</summary>
    private static bool GetAnswer(string question, string yes = "y", string no = "n") {
        Console.Write($"{question} ({yes}/{no})? ");
        string? answer = Console.ReadLine();
        return answer != null
            && string.Equals(answer.Trim(), yes, StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>Ask and return the desired board size.</summary>
    public static int GetBoardSize() {
        Console.Write($"Choose board size (default is {Settings.DefaultBoardSize}): ");
        string? input = Console.ReadLine();
        if (input != null && int.TryParse(input.Trim(), out int boardSize)) {
            if (boardSize is < Settings.MinBoardSize or > Settings.MaxBoardSize) {
                ColorPrint.PrintYellow(
                    $"Limiting board size to valid range {Settings.MinBoardSize}..{Settings.MaxBoardSize}"
                );
            }
            return Math.Clamp(boardSize, Settings.MinBoardSize, Settings.MaxBoardSize);
        }
        ColorPrint.PrintWarn($"Invalid size, defaulting to {Settings.DefaultBoardSize}...");
        return Settings.DefaultBoardSize;
    }
}

File: go/othello/othello.go
//==========================================================
// Othello
// Play Othello (Reversi) on the command line
// https://en.wikipedia.org/wiki/Reversi
// Akseli Lukkarila
// 2019-2026
//==========================================================

package othello

import (
	"fmt"
	"strings"
)

// Othello Gameplay loop and main logic.
type Othello struct {
	board        Board
	settings     Settings
	playerBlack  Player
	playerWhite  Player
	gameLog      []string
	gamesPlayed  int
	roundsPlayed int
}

// NewOthello Initialize Othello game.
func NewOthello(settings Settings) Othello {
	return Othello{
		board:        NewBoard(settings.BoardSize),
		settings:     settings,
		playerBlack:  *BlackPlayer(settings.ToPlayerSettings()),
		playerWhite:  *WhitePlayer(settings.ToPlayerSettings()),
		gameLog:      []string{},
		roundsPlayed: 0,
		gamesPlayed:  0,
	}
}

// Play one full game of Othello.
func (o *Othello) Play() {
	for {
		o.initGame()
		o.gameLoop()
		o.printResult()
		if o.settings.ShowLog {
			o.printLog()
		}
		if o.settings.AutoplayMode || !getAnswer("Would you like to play again", "y", "n") {
			break
		}
	}
}

// Initialize game board and players for a new game.
func (o *Othello) initGame() {
	// Re-use existing objects instead of initializing new ones
	if o.gamesPlayed > 0 {
		o.board = NewBoard(o.settings.BoardSize)
		o.playerBlack.Reset()
		o.playerWhite.Reset()
		o.roundsPlayed = 0
		o.gameLog = []string{}
	}
	if o.settings.AutoplayMode {
		// Computer plays both
		o.playerBlack.SetComputer()
		o.playerWhite.SetComputer()
	} else if o.settings.UseDefaults {
		// Default: play as black against white computer player
		o.playerWhite.SetComputer()
	} else if getAnswer("Would you like to play against the computer", "y", "n") {
		if getAnswer("Would you like to play as black or white", "b", "w") {
			o.playerWhite.SetComputer()
		} else {
			o.playerBlack.SetComputer()
		}
	}
	if !o.settings.CheckMode {
		PrintBold("\nPlayers:")
		o.printStatus()
	}
}

// Keep making moves until both players can't make a move any more.
func (o *Othello) gameLoop() {
	for o.board.CanPlay() && (o.playerBlack.CanPlay || o.playerWhite.CanPlay) {
		o.roundsPlayed++
		o.printRoundHeader()
		for _, player := range []*Player{&o.playerBlack, &o.playerWhite} {
			if result := player.PlayOneMove(&o.board); result != nil {
				o.gameLog = append(o.gameLog, fmt.Sprintf("%s;%s", *result, o.board.LogEntry()))
			}
			if !o.settings.CheckMode {
				fmt.Println("--------------------------------")
			}
		}
	}
	o.gamesPlayed++
	o.printGameEndFooter()
}

// Format game log with line numbers for each move.
func (o *Othello) formatGameLog() string {
	lines := make([]string, len(o.gameLog))
	for index, line := range o.gameLog {
		lines[index] = fmt.Sprintf("%02d: %s", index+1, line)
	}
	return strings.Join(lines, "\n")
}

// Print header for the current round.
func (o *Othello) printRoundHeader() {
	if !o.settings.CheckMode {
		PrintBold(fmt.Sprintf("\n=========== ROUND: %d ===========", o.roundsPlayed))
	}
}

// Print footer after the game has ended.
func (o *Othello) printGameEndFooter() {
	if !o.settings.CheckMode {
		PrintBold("\n================================")
		PrintGreenBold("The game is finished!\n")
	}
}

// Print game log which shows all moves made and the game board state after each move.
func (o *Othello) printLog() {
	formattedLog := o.formatGameLog()
	if !o.settings.CheckMode {
		PrintYellowBold("Game log:")
		fmt.Println(formattedLog)
	}
	hexHash := CalculateSHA256(formattedLog)
	fmt.Println(hexHash)
}

// Print ending status and winner info.
func (o *Othello) printResult() {
	PrintBold("Result:")
	o.printStatus()
	fmt.Println()

	winner := o.board.Result()
	if winner == Empty {
		fmt.Print("The game ended in a tie...\n\n")
	} else {
		fmt.Printf("The winner is %s!\n\n", winner.DiskString())
	}
}

// Print current board and player info.
func (o *Othello) printStatus() {
	fmt.Println(o.playerBlack.String())
	fmt.Printf("%s\n\n", o.playerWhite.String())
	fmt.Println(o.board.String())
}

// Ask a question with two options, and return bool from user answer.
func getAnswer(question, yes, no string) bool {
	fmt.Printf("%s (%s/%s)? ", question, yes, no)
	var input string
	if _, err := fmt.Scanln(&input); err == nil {
		return strings.ToLower(strings.TrimSpace(input)) == yes
	}
	return false
}

// GetBoardSize Ask and return the desired board size.
func GetBoardSize() int {
	fmt.Printf("Choose board size (default is %d): ", DefaultBoardSize)
	var input int
	if _, err := fmt.Scanf("%d", &input); err == nil {
		if input < MinBoardSize || input > MaxBoardSize {
			PrintYellow(fmt.Sprintf(
				"Limiting board size to valid range %d..%d", MinBoardSize, MaxBoardSize,
			))
		}
		return Clamp(input, MinBoardSize, MaxBoardSize)
	}
	PrintWarn("Invalid size, defaulting to %d...", DefaultBoardSize)
	return DefaultBoardSize
}
